src/Compilers/Core/Portable/ReferenceManager/CommonReferenceManager.Binding.cs

```diff
@@ -45,60 +57,224 @@ internal partial class CommonReferenceManager<TCompilation, TAssemblySymbol>
+                        {
+                            continue;
+                        }
+
+                        resolvedReferences.Add(peReference);
+                        resolvedMissingIdentities.Add(binding.ReferenceIdentity);
+                        implicitAssemblies.Add(data);
+
+                        var referenceBinding = data.BindAssemblyReferences(explicitAssemblies, IdentityComparer);
+                        referenceBindings.Add(referenceBinding);
+                        referenceBindingsToProcess.Push(referenceBinding);
+                    }
+                }
+
+                // NB: includes the assembly being built:
+                int explicitAssemblyCount = explicitAssemblies.Length;
+                allAssemblies = explicitAssemblies.AddRange(implicitAssemblies);
+                
+                // Rebind assembly referenced that were initially missing or were bound to a definitions with a different version 
+                // (with more assembly identities available we may have a better match now).
+
+                if (implicitAssemblies.Count > 0)
                 {
-                    return boundInputs;
+                    foreach (var referenceBinding in referenceBindings)
+                    {
+                        for (int i = 0; i < referenceBinding.Length; i++)
+                        {
+                            var binding = referenceBinding[i];
+
+                            if (!binding.IsBound || binding.VersionDifference != 0)
+                            {
+                                // We only need to resolve against implicitly resolved assemblies,
+                                // since we already resolved against explicitly specified ones.
+                                var newBinding = ResolveReferencedAssembly(
+                                    binding.ReferenceIdentity,
+                                    allAssemblies,
+                                    explicitAssemblyCount,
+                                    IdentityComparer);
+
+                                if (IsBetterVersionMatch(binding.ReferenceIdentity, allAssemblies, binding, newBinding))
+                                {
+                                    referenceBinding[i] = newBinding;
+                                }
+                            }
+                        }
+                    }
+
+                    UpdateBindingsOfAssemblyBeingBuilt(referenceBindings, explicitAssemblies, implicitAssemblies);
                 }
             }
+            finally
+            {
+                implicitAssemblies.Free();
+                resolvedMissingIdentities.Free();
+                referenceBindingsToProcess.Free();
+            }
+        }
+
+        private static void UpdateBindingsOfAssemblyBeingBuilt(ArrayBuilder<AssemblyReferenceBinding[]> referenceBindings, ImmutableArray<AssemblyData> explicitAssemblies, ArrayBuilder<AssemblyData> implicitAssemblies)
```

Comment:
Consider passing a number of explicit references (i.e. explicitAssemblies.Lenght -1) instead of explicitAssemblies array.